numpy/__init__.cython-30.pxd

```diff
@@ -657,7 +657,7 @@ cdef extern from "numpy/arrayobject.h":
     #object PyArray_FromArrayAttr (object, dtype, object)
     #NPY_SCALARKIND PyArray_ScalarKind (int, ndarray*)
     int PyArray_CanCoerceScalar (int, int, NPY_SCALARKIND)
-    object PyArray_NewFlagsObject (object)
+    object PyArray_NewFlagsObject (object, object)
```

Comment:
It might be interesting to see which functions would need somthing like this.  I am removing this one in gh-25292.
But more importantly: you must not change this, it is public API/ABI.  We can change public API in a way that allows this right now (i.e. with a runtime check that drops passing the module), because a major release is upcoming and we are forcing users to recompile.  But assuming we won't pull all of them off, it will be more useful to finalize a transition in some future.

What is necessary, is a way to do this without changing anything.  Maybe by introducing a second table/entry-point.  Or maybe we even can pass something (like even `NULL`) right now based on the NumPy version which could be the module in the future.
But, I am not sure how much it matters to allow backcompat with old versions, so just introducing new functions might be just as well.  (And overall, you can always pull a off a compatibility mode if you add new API, it's just a bit ugly and you may require *compiling* against the newest version.)

Or, you start without touching public API for now.

Reply:
@seberg I could attempt an audit to see what API changes are necessary for heap types, but it's hard to guarantee without actually doing the migration to heap types. What is absolutely certain is that all of the global type structs are not compatible with heap types.  In this example it is `PyArrayFlags_Type` which is now in slot 4 of `multiarray_types_api`. The entirety of `multiarray_types_api` is not compatible with heap types.

If I understand you correctly, do you mean to say that you want to _add_ all the APIs that will be needed for heap types, and then implement them correctly with the current static types. Then, the APIs that are impossible to serve when the types are migrated to heap types will be marked as obsolete.  And then at some point in the future, you'll remove the obsolete APIs when or before the actual migration to heap types take place. Do I understand this correctly?

Reply:
> the APIs that are impossible to serve when the types are migrated to heap types will be marked as obsolete

But, the APIs shouldn't be impossible, right?  They might not support subinterpreters, or be slow when using them, but they should remain possible, I think!?

> you'll remove the obsolete APIs when or before the actual migration to heap types take place

I don't think there is anything preventing us from using heap types?  Sure, we still either need the global version of the types, but they might just exist for the sake of the *public* API.  A public API may not support subinterpreters.

> audit to see what API changes are necessary for heap types

I was just curious, because maybe one could implement the public API changes even in the right now.  Chances are that isn't worthwhile, though.  We will probably need a lot of macros like:
```
def C_API_NEW(...):
    if numpy_version > x.y:
        C_API_WITH_MOD(mod, ...)
    else:
        OLD_C_API(...)
```
or a similar pattern with `import_array_with_mod()`, that inserts some such aliases to a new table.
In principle, we could even export a different table for each subinterpreter if that makes things much easier as that goes through `getattr(numpy._core._multiarray_umath.ARRAY_API)` (although I admit that most of the tables would always be the same functions).

---

In either case, long story short. I don't see why we cannot just migrate to heap-types.  *Downstream* may not be able to support subinterpreters without new API, though.  And even when we add that new API, we will need hacks, because almost any downstream needs to support 3 year old NumPy.  That can be hacked with macros as above.  If we would be able to figure that out in a month for 2.0, that might make it a tiny bit easier, but I don't think enough that we should worry about it.

Reply:
@seberg 

consider this scenario. 

The change to use heap types looks would look like this ( see line 4924 [here](https://github.com/numpy/numpy/pull/25359/files#diff-71aadf0228c9cd61ea260748af1dcd4a726eea8476b8cbd667367466c1e39077)):
```
- PyType_Ready(&PyArrayFlags_Type)
+ state->arrayflags_type = (PyTypeObject *)PyType_FromModuleAndSpec(m, &arrayflags_type_spec, NULL);
```
numpy makes the above change, but continues to provide the current API in addition to the new API. A user has the following function implemented:

```
#include "numpy.h"
static int is_arrayflags_type(PyObject* mod, PyObject* obj) {
    return PyObject_TypeCheck(obj, &PyArrayFlags_Type);
}
```

If numpy provides the same API, but changes to use heap types, the above code will indeed compile, but at runtime, it will always return 0, whether or not `obj` is an instance of the flags type. The same is true for all types in the `multiarray_types_api` array. Whether or not a subinterpreter is used is irrelevant, the module exec function will add the type using PyType. This is because all instance of the ArrayFlags type, and indeed any type, will have been created based on the new type spec, and not on the global static type. The implementation of PyObject_TypeCheck turns into nothing but an identity check on the pointer itself. See [here](https://github.com/python/cpython/blob/1c5fc02fd0576be125638a5261be12eb3224be81/Include/object.h#L351)

Reply:
> but at runtime, it will always return 0

It will work just fine, so long you are not using the public C-API in subinterpreters.  I don't think anything stops us from using heap types *internally*, the main issue is getting it right for the public API to allow e.g. SciPy to also support subinterpreters.
(And untangling the DType mess, since Python used to be buggy around extension heap types.)

In some places, maybe you can actually implement things by just finding the module state without it being passed in, since I think that should be possible, although maybe a bit slow.

That requires some new API clearly (downstream needs to store and pass the module state).  If there is enough clarity, maybe we could already force some of those changes that are necessary right now.  If not, I assume we can add those changes as additional API at any time (although backporting may be annoying if projects care about running on older versions of NumPy).

Reply:
It's also very late in the game to get these sorts of change in for NumPy 2.0. Of course, maybe we'll need a NumPy 3.0 to properly support nogil python or subinterpreters and we can do those changes then. It's going to be a hard sell IMO to make this level of change before 2.0 gets cut when we're still working on getting the changes we've already planned done.

Reply:
Right, basically italwaysseemed to me likewise need new api anyway and thus it's not a major issue to miss 2.0

Reply:
OK, but to get back on track.  The pragmatic thing I think is to start with changing to heap types internally. The public app table would only be populated once and be global state, not subinterpreter ready.

Reply:
@seberg I think I understand your proposal, I plan to submit another PR that attempts to implement it using this arrayflags type.

I see two risks:
1. The users of the numpy C API do not control whether subinterpreters are or are not used, and so I don't know if the "failure mode" will be sensible or something terrible like undefined behavior.
2. The current API exposes static types, I don't know exactly what it would mean for the first heap type to be the same as the static type, since the static type will be stored in the .rodata section of the numpy shared object, while the heap type will be on the heap. I don't know how that could be made to pass the PyObject_TypeCheck implementation.

Nonetheless I will give it a shot

Reply:
> The users of the numpy C API do not control whether subinterpreters

Yes, they do.  Users of the C-API are extension modules, which have to opt in for (proper) subinterpreter support.  They cannot *support* them which is too bad maybe, but there is no control issue.

> The current API exposes static types

It doesn't matter at all whether the API exposes static or heap types.